teamspeak3: modernise

Atemu opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Atemu committed:
teamspeak_client: don't wrap with cc's libdir

This pulls in the entire C compiler into the runtime closure. It works just fine
without this.

Atemu committed:
teamspeak_client: use regular libcxx

No reason to use the one from LLVM.

Atemu committed:
teamspeak_client: refactor QT deps

This is the modern method of accessing them and is compatible with by-name.

Atemu committed:
teamspeak_client: run phase hooks

Atemu committed:
teamspeak_client: make libredirect a regular runtimeDep

I don't see a need to preload here

Atemu committed:
teamspeak_client: use wrapQtAppsHook

Atemu committed:
teamspeak_client: refactor libquazip patching

Atemu committed:
teamspeak_client: use autoPatchelfHook rather than manual patchelf

Atemu committed:
teamspeak_client: remove unnecessary dependencies

I honestly don't know what these are here for but they're not in any of the
binaries' NEEDED sections, so let's yeet them.

autoPatchelfHook ensures we have all required libraries.

Atemu committed:
teamspeak_client: modernise installPhase

Atemu committed:
teamspeak_client: install to opt/ subdirectory

This is not a library.

This actually matters somewhat because lib/ is installed into the system-wide
path on NixOS for whatever reason.

Atemu committed:
teamspeak_client: run installer script without -x by default

It's only relevant for debugging the package when updating.

Atemu committed:
teamspeak3: rename from teamspeak_client

This is what people would be looking for when searching for "teamspeak".

Barely anyone will need the server and those who do will likely explicitly look
for the server rather than just "teamspeak".

teamspeak_server should probably be renamed to teamspeak3-server.

Atemu commented on pkgs/by-name/te/teamspeak3/package.nix:
```diff
@@ -0,0 +1,143 @@
+  ];
+
+  qtWrapperArgs = [
+    # wayland is currently broken, remove when TS3 fixes that
+    "--set QT_QPA_PLATFORM xcb"
```
I checked and it's indeed still necessary.

ewancg replied:
I don't know of an appropriate place to make the issue

Scrumplex commented on pkgs/by-name/te/teamspeak3/package.nix:
```diff
@@ -0,0 +1,143 @@
+in
+
+stdenv.mkDerivation rec {
+  pname = "teamspeak-client";
```
```suggestion
  pname = "teamspeak3";
```

Atemu replied:
Ah right, we had that convention.

Scrumplex commented on pkgs/by-name/te/teamspeak3/package.nix:
```diff
@@ -0,0 +1,143 @@
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    description = "TeamSpeak voice communication tool";
+    homepage = "https://teamspeak.com/";
+    sourceProvenance = with sourceTypes; [ binaryNativeCode ];
+    license = {
+      # See distribution-permit.txt for a confirmation that nixpkgs is allowed to distribute TeamSpeak.
+      fullName = "Teamspeak client license";
+      url = "https://www.teamspeak.com/en/privacy-and-terms/";
+      free = false;
+    };
+    maintainers = with maintainers; [
+      lhvwb
+      lukegb
+      atemu
+    ];
+    mainProgram = "ts3client";
+    platforms = [ "x86_64-linux" ];
+  };
```
nit
```suggestion
  meta = {
    description = "TeamSpeak voice communication tool";
    homepage = "https://teamspeak.com/";
    sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ];
    license = {
      # See distribution-permit.txt for a confirmation that nixpkgs is allowed to distribute TeamSpeak.
      fullName = "Teamspeak client license";
      url = "https://www.teamspeak.com/en/privacy-and-terms/";
      free = false;
    };
    maintainers = with lib.maintainers; [
      lhvwb
      lukegb
      atemu
    ];
    mainProgram = "ts3client";
    platforms = [ "x86_64-linux" ];
  };
```

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/teamspeak/client.nix:
```diff
@@ -141,14 +142,16 @@ stdenv.mkDerivation rec {
+
+  preFixup =
+    ''
       wrapProgram $out/bin/ts3client \
```
Are we doing now double wrapping here?

Atemu replied:
I solved that particular problem in a later commit. One step at a time.

lukegb commented on pkgs/by-name/te/teamspeak3/package.nix:
```diff
@@ -0,0 +1,143 @@
+}:
+
+let
+  arch = "amd64";
```
I don't think this is useful; we parameterise source from it, but there's no other build right now, and there's a handful of other places we reference the arch (e.g. when going into the directory once we've unpacked the installer).

Atemu replied:
I just retained this from the original drv as I didn't have a strong opinion on this and saw it as fine as is.

lukegb commented on pkgs/by-name/te/teamspeak3/package.nix:
```diff
@@ -0,0 +1,143 @@
+  buildPhase = ''
+    runHook preBuild
+
+    # Run the installer script non-interactively
+    echo -e '\ny' | PAGER=cat sh -e $src
+
+    # Delete unecessary libraries - these are provided by nixos.
+    pushd TeamSpeak3-Client-linux_amd64/
```
I'm not really a huge fan of this. Doing this in unpackPhase means it's somewhat expected to have gone into the subdirectory that this generates at the end anyway (and probably should be specified as sourceRoot, for instance), so the `mv * ..` isn't required.

Atemu replied:
The thing is that this needs to happen after the installer has run. It seems more sensible to me to have the installer be ran in the `buildPhase` rather than the `unpackPhase` as that's the primary process that produces the desired binaries.

I now have the directory switching encapsulated within the buildPhase though which I also like better.

Scrumplex replied:
> It seems more sensible to me to have the installer be ran in the `buildPhase` rather than the `unpackPhase` as that's the primary process that produces the desired binaries.

I'd argue that we are not producing binaries, but merely extracting them. The same way we let `dpkg` extract .deb files in unpackPhase, I'd argree with @lukegb, that we should extract these files durin unpackPhase. That would also allow people to add `patches` and other tasks to the patch phases.

lukegb commented on pkgs/by-name/te/teamspeak3/package.nix:
```diff
@@ -0,0 +1,143 @@
+    pushd TeamSpeak3-Client-linux_amd64/
+
+    # I hate `find`.
+    find . -\( -name '*.so' -or -name '*.so.*' -\) -print0 | xargs -0 rm
```
Why `-print0 | xargs -0 rm` vs `-delete`?

Atemu replied:
I invite you to try it; it does not work. `find` is historical garbage and even getting the print0 to work wasn't trivial. I was tempted to just use `fd` here.

lukegb replied:
I tried it; it seems to work?

```
$ rg find pkgs/by-name/te/teamspeak3/package.nix 
60:  # it's looking for a dependency and didn't find it. Check the script and make
77:    find . -\( -name '*.so' -or -name '*.so.*' -\) -delete
$ NIXPKGS_ALLOW_UNFREE=1 nix-build -A teamspeak3 --option builders ''
/nix/store/hx12r49pbb533in5yvz96qq17a4f54vf-teamspeak3-3.6.2
$ ls -R /nix/store/hx12r49pbb533in5yvz96qq17a4f54vf-teamspeak3-3.6.2 | rg '\.so'
$
```

Is there something else that should be being deleted here?

lukegb replied:
I really don't care about this as much as the other thing though, to be honest, so whatever.

lukegb reviewed (commented):
Thanks for the cleanups! I have a few minor comments, but on the whole thanks for the cleanup.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/teamspeak/client.nix:
```diff
@@ -90,34 +62,65 @@ stdenv.mkDerivation rec {
 
-  buildPhase = ''
-    runHook preBuild
+    # I hate `find`.
```
```suggestion
    # Delete unecessary libraries - these are provided by nixos
```
me, too

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/teamspeak/client.nix:
```diff
@@ -90,34 +62,65 @@ stdenv.mkDerivation rec {
+    # I hate `find`.
+    find . -\( -name '*.so' -or -name '*.so.*' -\) -print0 | xargs -0 rm
+    rm QtWebEngineProcess
+    rm qt.conf
```
```suggestion
    rm QtWebEngineProcess qt.conf
```

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/teamspeak/client.nix:
```diff
@@ -90,34 +62,65 @@ stdenv.mkDerivation rec {
 
-    runHook postUnpack
-  '';
+    # Delete unecessary libraries - these are provided by nixos.
```
```suggestion
```

Atemu replied:
This is precisely where it's intended to be.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/teamspeak/client.nix:
```diff
@@ -97,34 +84,38 @@ stdenv.mkDerivation rec {
+  qtWrapperArgs = [
+    # wayland is currently broken, remove when TS3 fixes that
+    "--set QT_QPA_PLATFORM xcb"
+    "--set NIX_REDIRECTS /usr/share/X11/xkb=${xkeyboard_config}/share/X11/xkb"
```
Didn't we remove the libredirect earlier?

Atemu replied:
I was not aware that libredirect is a Nixpkgs-specific thing. The application appears to be working fine without it to me. How would this manifest?

Atemu commented on pkgs/applications/networking/instant-messengers/teamspeak/client.nix:
```diff
@@ -49,6 +49,7 @@ let
     libsForQt5.quazip
     libcxx
     libredirect
+    libsForQt5.quazip
```
Rebase error, thanks.

Atemu committed:
teamspeak3: remove NIX_REDIRECTS

The lib was removed earlier, so this doen't have any effect. It didn't appear to
cause any issues.

Atemu committed:
teamspeak3: avoid `with lib;`

Shawn8901 commented on pkgs/by-name/te/teamspeak3/package.nix:
As there is also a teamspeak3 server in nixpkgs (currently called teamspeak_server) should we keep some indication in the package name that this is the client?

Atemu replied:
I elaborated on this in the commit message.

Shawn8901 replied:
Yeah okay when we at some point then rename the server (no need to be done here), I think I am okay

Atemu replied:
I don't maintain the server and have never used it, so I don't see that as part of my responsibility.

Shawn8901 replied:
yeah also i did not try to say that, so all good :slightly_smiling_face:

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/help-making-a-derivation-fixed/55715/4

Scrumplex reviewed (approved):
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 354161`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>teamspeak3</li>
  </ul>
</details>

lukegb commented on pkgs/by-name/te/teamspeak3/package.nix:
```diff
@@ -0,0 +1,143 @@
+    # Run the installer script non-interactively
+    echo -e '\ny' | PAGER=cat sh -e $src
+
+    mv TeamSpeak3-Client-linux_${arch}/* .
```
You can just cd into this. It is expected (even for stdenv unpackPhase) that you are in the unpacked content after unpackPhase has completed.

(https://nixos.org/manual/nixpkgs/stable/#var-stdenv-sourceRoot)

Atemu replied:
I can see where you're coming from but I think it's just a lot simpler to do it this way.

lukegb replied:
I think it would be helpful to justify why you think this is simpler - it's possible I'm missing something here.

My argument is that this approach is different to other unpackPhases (and in particular the stock stdenv one), which are usually "take $src file, uncompress it, and change the working directory into the decompressed content". In general, I would consider it simpler to be as similar as possible to other derivations.

Atemu replied:
I don't think it's worth discussing this at this point tbh. This is getting into bikeshedding territory.

lukegb reviewed (approved):
One minor comment but otherwise LG.

Shawn8901 commented:
Not yet sure why, so maybe someone else also had/has the problem. After cherry-picking the changes and applying everything Teamspeak3 no longer shows up in the Start Menu. I installed it via `home.packages` with home-manager.

Atemu commented:
I'll have to check that the desktop item is unchanged. Some start menus require a restart for new nix desktop items to become available, have you done that?

Shawn8901 commented:
yeah i noticed it yesterday and it did not came up today. I also can not see on first glance the issue tbh, like you said that should be unchanged.

Atemu commented:
Well, the desktop item isn't there....

I forgot you need to add a hook to do that.

lukegb committed:
teamspeak3: drop 'arch' variable

This variable is always set to "amd64", so there's no point keeping it.
If there's an aarch64 version in future it can be added back.

Shawn8901 reviewed (approved):
shows up again :slightly_smiling_face: . I was that far that i noticed that the .desktop file is missing on my own checks.

lgtm now
